framework/validators/EmailValidator.php

```diff
@@ -111,21 +112,29 @@ protected function validateValue($value)
      */
     protected function isDNSValid($domain)
     {
-        if (checkdnsrr($domain . '.', 'MX')) {
-            $mxRecords = dns_get_record($domain . '.', DNS_MX);
-            if ($mxRecords !== false && count($mxRecords) > 0) {
-                return true;
-            }
+        if ($this->hasDNSRecord($domain, true)) {
+            return true;
         }
 
-        if (checkdnsrr($domain . '.', 'A')) {
-            $aRecords = dns_get_record($domain . '.', DNS_A);
-            if ($aRecords !== false && count($aRecords) > 0) {
-                return true;
-            }
+        return $this->hasDNSRecord($domain, false) === true;
+    }
+
+    private function hasDNSRecord($domain, $mx)
+    {
+        $normalizedDomain = $domain . '.';
+        if (!checkdnsrr($normalizedDomain, ($mx ? 'MX' : 'A'))) {
+            return false;
+        }
+
+        try {
+            // dns_get_record can return false and emit Warning that may or may not be converted to ErrorException
+            $records = dns_get_record($normalizedDomain, ($mx ? DNS_MX : DNS_A));
+        } catch (ErrorException $exception) {
+            Yii::warning($exception->getMessage(). '. Domain: "' . $domain . '"', __METHOD__);
```

Comment:
Why warning? If this exception will be thrown every time when user provide invalid domain, then it shouldn't be a warning - handling garbage input is purpose of this validator, there is nothing special in this situation.

Reply:
Adding warning/error was a suggestion by [samdark](https://github.com/yiisoft/yii2/issues/17602#issuecomment-539504117) and [kamarton](https://github.com/yiisoft/yii2/issues/17602#issuecomment-539885255). 
At first I was going to avoid emitting warnings but after catching `dns_get_record(): Unable to parse DNS data received.` I'm not sure what's better - to discard error silently or to notify developer.

Reply:
Logging could be limited to `Yii::debug()`. But I personally would just ignore it - warning is expected here. Handling this warning is internal logic of this validator, it should not pollute application logs - this is nothing that end user should worry about.

Reply:
Removed warnings as soon as site owner can do nothing to fix malformed replies from external DNS server.

Reply:
> Logging could be limited to Yii::debug()...

@rob006 I think this is a bug that the developer or hostmasters should be aware of. If they cannot resolve it, they have the option to deactivate this feature. If the error is absorbed, the return value is false, which produces unwanted results without feedback.

The [`dns_get_record()`](https://www.php.net/manual/en/function.dns-get-record.php) documentation does not documented an exception at all, so the presence of an exception indicates an error.

Reply:
`dns_get_record()` check is workaround for Alpine bug - every other environment will use `checkdnsrr()`, which returns `false` without any warning. Ignoring warning here just makes this behavior consistent on all platforms.

> which produces unwanted results without feedback

Returning `false` is expected result for domain with invalid DNS zone settings - it is unreachable anyway.